dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,278 @@
+    private const string GetOperationName = "Get";
+
+    private readonly Sdk.PineconeClient _pineconeClient;
+    private readonly PineconeVectorStoreRecordCollectionOptions<TRecord> _options;
+    private readonly IVectorStoreRecordMapper<TRecord, Sdk.Vector> _mapper;
+
+    /// <summary>
+    /// Map from an index name to the corresponding Pinecone.NET Index object.
+    /// </summary>
+    private readonly ConcurrentDictionary<string, Sdk.Index<GrpcTransport>> _indexMap = new();
+
+    /// <inheritdoc />
+    public string CollectionName { get; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PineconeVectorStoreRecordCollection{TRecord}"/> class.
+    /// </summary>
+    /// <param name="pineconeClient">Pinecone client that can be used to manage the collections and vectors in a Pinecone store.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    /// <exception cref="ArgumentNullException">Thrown if the <paramref name="pineconeClient"/> is null.</exception>
+    /// <param name="collectionName">The name of the collection that this <see cref="PineconeVectorStoreRecordCollection{TRecord}"/> will access.</param>
+    /// <exception cref="ArgumentException">Thrown for any misconfigured options.</exception>
+    public PineconeVectorStoreRecordCollection(Sdk.PineconeClient pineconeClient, string collectionName, PineconeVectorStoreRecordCollectionOptions<TRecord>? options = null)
+    {
+        Verify.NotNull(pineconeClient);
+
+        this._pineconeClient = pineconeClient;
+        this.CollectionName = collectionName;
+        this._options = options ?? new PineconeVectorStoreRecordCollectionOptions<TRecord>();
+
+        if (this._options.MapperType == PineconeRecordMapperType.PineconeVectorCustomMapper)
+        {
+            if (this._options.VectorCustomMapper is null)
+            {
+                throw new ArgumentException($"The {nameof(PineconeVectorStoreRecordCollectionOptions<TRecord>.VectorCustomMapper)} option needs to be set if a {nameof(PineconeVectorStoreRecordCollectionOptions<TRecord>.MapperType)} of {nameof(PineconeRecordMapperType.PineconeVectorCustomMapper)} has been chosen.", nameof(options));
+            }
+
+            this._mapper = this._options.VectorCustomMapper;
+        }
+        else
+        {
+            this._mapper = new PineconeVectorStoreRecordMapper<TRecord>(new PineconeVectorStoreRecordMapperOptions
+            {
+                VectorStoreRecordDefinition = this._options.VectorStoreRecordDefinition
+            });
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<bool> CollectionExistsAsync(CancellationToken cancellationToken = default)
+    {
+        var result = await this.RunOperationAsync(
+            CollectionExistsName,
+            async () =>
+            {
+                var collections = await this._pineconeClient.ListIndexes(cancellationToken).ConfigureAwait(false);
```

Comment:
Is it possible to use `_pineconeClient.GetIndex` to check for the existence of an index?  It might be more scalable if there are many indexes and listing them all becomes problematic.

Reply:
`GetIndex` is a heavier operation that creates an index instance including retrieving all the data about it and instantiating the underlying transport, and in the negative case it throws. It is going to be more expensive than retrieving an array of index details and scanning it😅 The breakout point in cost is going to be at least dozens or hundreds of indexes...

Reply:
My main concern was that GetIndex throws if index doesn't exist, as @neon-sunset pointed out, so we would have to do error handling. I'm not a fan of gobbling up exceptions and that's probably what we'd have to do, no? Checking for the existence of an index that is not there is a normal thing to do, but logs (when they are implemented) would indicate errors.

Reply:
Also, not sure if the existence check should take advantage of cache or rather always go for the source of truth instead. My thinking was that if user is actually calling this method (as opposed to just blindly calling a get/upsert/delete operation), they have a suspicion that the index might not be there after all, or at least they want to really be sure the index is there. Example where it would make a difference is when user the index was there before and we used it, but then it got deleted outside SK (e.g. using Pinecone.NET SDK directly or from the pinecone website). Thoughts on this, wrt SK contract for this method? @westey-m @dmytrostruk

Reply:
Spoke with @westey-m and we agreed on skipping cache and always querying the underlying database.